test_conformance/subgroups/subhelpers.h

```diff
@@ -354,25 +1231,55 @@ struct test
         clProgramWrapper program;
         clKernelWrapper kernel;
         cl_platform_id platform;
-        cl_int sgmap[2 * GSIZE];
+        cl_int sgmap[4 * GSIZE];
         Ty mapin[LSIZE];
         Ty mapout[LSIZE];
+        std::stringstream kernel_sstr;
 
+        kernel_sstr << "#define NON_UNIFORM_WG_SIZE ";
+        kernel_sstr << NON_UNIFORM_WG_SIZE << "\n";
         // Make sure a test of type Ty is supported by the device
-        if (!TypeCheck<Ty>::val(device)) return 0;
+        if (!TypeManager<Ty>::type_supported(device))
+        {
+            log_info("Data type not supported : %s\n", TypeManager<Ty>::name());
+            return 0;
+        }
+        else
+        {
+            if (strstr(TypeManager<Ty>::name(), "double"))
+            {
+                kernel_sstr << "#pragma OPENCL EXTENSION cl_khr_fp64: enable\n";
+            }
+            else if (strstr(TypeManager<Ty>::name(), "half"))
+            {
+                kernel_sstr << "#pragma OPENCL EXTENSION cl_khr_fp16: enable\n";
+            }
+        }
+
+        for (std::string extension : required_extensions)
+        {
+            if (!is_extension_available(device, extension.c_str()))
+            {
+                log_info("The extension %s not supported on this device. SKIP "
+                         "testing - kernel %s data type %s\n",
+                         extension.c_str(), kname, TypeManager<Ty>::name());
+                return 0;
+            }
+            kernel_sstr << "#pragma OPENCL EXTENSION " + extension
```

Comment:
The extended subgroups don't require the pragmas, so I think this line should be removed!

Reply:
FYI I don't think we need the pragmas for cl_khr_subgroups either for the functionality being tested by the PR.

Reply:
I agree, these should not be necessary as the spec doesn't say they are necessary. I suppose that this change can be done later, once https://github.com/KhronosGroup/OpenCL-Docs/issues/82 is resolved for example.

Aside from that, there are currently three ways to introduce `#pragma` in the kernel used by these tests:
1. When `doublen` or `halfn` are used.
2. When `useCoreSubgroups` is true.
3. When `required_extensions` is non-empty.

In other words, this cleanup would simplify the code in this function as only 1. would be necessary.